acceptable/__main__.py

```diff
@@ -0,0 +1,118 @@
+import argparse
+import json
+from pathlib import Path
+import sys
+
+from jinja2 import Environment, PackageLoader
+import yaml
+
+from acceptable._build_doubles import extract_schemas_from_file
+
+env = Environment(
+    loader=PackageLoader('acceptable', 'templates'),
+    autoescape=False,
+)
+
+
+def main():
+    cli_args = parse_args()
+    sys.exit(cli_args.func(cli_args))
+
+
+def parse_args(raw_args=None, parser_cls=None, stdin=None):
+    if raw_args is None:
+        raw_args = sys.argv[1:]
+    if parser_cls is None:
+        parser_cls = argparse.ArgumentParser
+    if stdin is None:
+        stdin = sys.stdin
+
+    parser = parser_cls(
+        description='Tool for working with acceptable metadata',
+    )
+    subparser = parser.add_subparsers(dest='cmd')
+    subparser.required = True
+
+    metadata_parser = subparser.add_parser(
+        'metadata', help='Scan files and print extracted metadata in json')
+    metadata_parser.add_argument('files', nargs='+')
+    metadata_parser.set_defaults(func=metadata_cmd)
+
+    render_parser = subparser.add_parser(
+        'render', help='Render markdown documentation for api metadata'
+    )
+    render_parser.add_argument(
+        'metadata',
+        nargs='?',
+        type=argparse.FileType('r'),
+        default=stdin,
+    )
+    render_parser.add_argument(
+        '--name', '-n', required=True, help='Name of service')
+    render_parser.add_argument(
+        '--dir', '-d', default='docs', help='output directory')
+    render_parser.set_defaults(func=render_cmd)
+
+    return parser.parse_args(raw_args)
+
+
+def metadata_cmd(cli_args):
+    metadata = scan_metadata(cli_args.files)
+    print(json.dumps(metadata, indent=2, sort_keys=True))
+    return 0
+
+
+def scan_metadata(files):
+    # This uses AST parsing for now. In future, we might just import code.
+    schemas = []
+    metadata = {}
+    for path in files:
+        schemas.extend(extract_schemas_from_file(path))
+    for schema in schemas:
+        metadata[schema.view_name] = {
+            'api_name': schema.view_name,
+            'version': int(float(schema.version)),  # convert any '1.0' strings
+            'methods': schema.methods,
+            'url': schema.url,
+            'request_schema': schema.input_schema,
+            'response_schema': schema.output_schema,
+            'doc': schema.doc,
+        }
+    return metadata
+
+
+def render_cmd(cli_args):
+    root_dir = Path(cli_args.dir)
+    root_dir.joinpath('en').mkdir(parents=True, exist_ok=True)
+    metadata = json.load(cli_args.metadata)
```

Comment:
As this is a developer tool, I thought perhaps the stack trace was the best error handling. The error message usually points to the line/column that doesn't parse?

Reply:
Well, I'll leave this up to you (since it's your code now anyway), but I don't buy the "traceback is the best error report for developers" argument. Personally, I'd *much* rather see a well formatted error message with the file, line:col in the message.